### PR TITLE
8324950: IGV: save the state to a file

### PR DESCRIPTION
**Pull request by tobiasholenstein**
The current workflow in IGV is the following:
1) import an XML file with graphs or send via network
2) open or more graphs in a tab
3) extract a set of nodes to be displayed in the tab
4) close IGV and start from 1) again

The idea of this RFE is to save the opened graph tabs and extracted nodes of a graph in the `graph.xml` file.
### The new workflow
 
When opening IGV the user gets an empty workspace without any opened files. 
- Graphs can be sent via the network to IGV
- Graph can be opened from an XML file
empty

Unzipping this [example_graph.zip](https://github.com/openjdk/jdk/files/14946834/example_graph.zip) and opening `graphs.xml` shows the following graph. New with this RFE is that opened graph tabs and extracted nodes are saved to the  `graph.xml` file and restored when re-opening the `graphs.xml`:

graph

A new `<graphStates>` is introduced in `graphs.xml` that stores the opened graphs and their visible nodes: 
```xml
<graphDocument> <!-- "graphs.xml" -->
    <group> <!-- "1 - Group" -->
        <graph> <!-- "1. Graph" -->
            <properties>...</properties>
            <nodes>...</nodes>
            <edges>...</edges>
            <controlFlow>...</controlFlow>
            <graphStates> <!-- All the tabs that are open for a graph -->
                <state> <!-- a single open tab -->
                    <difference value="0"/> <!-- larger zero are difference graphs -->
                    <visibleNodes all="false">
                        <node id="0"/>
                        <node id="1"/>
                        <node id="3"/>
                    </visibleNodes >
                </state>
            </graphStates>
        </graph>
    </group>
</graphDocument>
```

The workspace menu is restructured:

open

- `Open` allows the user to open an XML file. In IGV there is either no XML opened indicated as `untitled` or exactly one xml file opened. It's not possible to have two XML files opened at the same time: 
- `Import`:  Allows the user to import graphs from another XML file to the current opened XML file.

save

- `Save..` saves the current opened xml file. Create a new file if no file is opened. 
- `Save as...` save the current graphs as a copy to an xml file. 
Note: there is no autosave and IGV also does not ask if you want to save changes when closing it. 

remove

- `Remove selected graphs and groups`
- `clear workspace`: closes all tabs, removes all groups/graphs and closes the XML file. Does NOT autosave.

### Implementation

Actions classes are refactored to call a function in OutlineTopComponent:

`OpenAction.java` calls `OutlineTopComponent.openFile()`: Opens a file dialog to select and load a graph document. Clears the workspace and adds the loaded document to the workspace. Loads a graph document from the given file path, updating progress via a ProgressHandle. Parses the XML file, adds the parsed document to the workspace, and loads associated contexts if specified.

`SaveAction.java` calls `OutlineTopComponent.save()`: Saves the current graph document. If the document has no location, call saveAs().

`SaveAsAction.java` calls `OutlineTopComponent.saveAs()`: Stores the provided graph document to the designated file path, optionally with associated contexts.

`ExportAction.java` calls `OutlineTopComponent.exportToXML(GraphDocument doc)`: Exports the provided graph document to an XML file. Displays a file chooser dialog to specify the location and name of the XML file.

`ImportAction.java` calls `OutlineTopComponent.importFromXML()`: Imports graph documents from one or more XML files. Displays a file chooser dialog to select one or multiple XML files for import. Each selected file is added to the workspace.

`RemoveAllAction.java` calls `OutlineTopComponent.clearWorkspace()`: Clears the workspace by resetting the document path, clearing the document, and resetting the folder structure. After clearing the workspace, it will be ready for new documents.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324950](https://bugs.openjdk.org/browse/JDK-8324950): IGV: save the state opened graphs to a file (**Enhancement** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17630/head:pull/17630` \
`$ git checkout pull/17630`

Update a local copy of the PR: \
`$ git checkout pull/17630` \
`$ git pull https://git.openjdk.org/jdk.git pull/17630/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17630`

View PR using the GUI difftool: \
`$ git pr show -t 17630`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17630.diff">https://git.openjdk.org/jdk/pull/17630.diff</a>

</details>

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324950](https://bugs.openjdk.org/browse/JDK-8324950): IGV: save the state to a file (**Enhancement** - P4)

### Reviewers
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**) ⚠️ Review applies to [a4b48c43](https://git.openjdk.org/jdk/pull/17630/files/a4b48c436704644a1ca57c90a323da0a20db4ef9)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17630/head:pull/17630` \
`$ git checkout pull/17630`

Update a local copy of the PR: \
`$ git checkout pull/17630` \
`$ git pull https://git.openjdk.org/jdk.git pull/17630/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17630`

View PR using the GUI difftool: \
`$ git pr show -t 17630`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17630.diff">https://git.openjdk.org/jdk/pull/17630.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17630#issuecomment-1948357743)

### CODE REVIEW

**Review by TobiHartmann — COMMENTED** [review]
Very nice. I gave this a first try and it works well on my machine (Ubuntu, x64).

File: src/utils/IdealGraphVisualizer/igv.sh [hunk 1]
```diff
@@ -20,4 +20,13 @@
 # or visit www.oracle.com if you need additional information or have any
 # questions.
 
-mvn --batch-mode -f application/pom.xml nbm:run-platform >.igv.log 2>&1
+DEFAULT_WORKSPACE="$(pwd)/workspace"
+DEFAULT_WORKSPACE=${IGV_WORKSPACE:-$DEFAULT_WORKSPACE}
+WORKSPACE=${1:-$DEFAULT_WORKSPACE}
+echo "The IGV workspace in which the imported graphs are saved saved can be set as follows :"
```
Comment (robcasloz) [inline src/utils/IdealGraphVisualizer/igv.sh:26]:
```suggestion
echo "The IGV workspace in which the imported graphs are saved can be set as follows :"
```

**Review by robcasloz — CHANGES_REQUESTED** [review]
Nice idea!
When I start IGV with `sh igv.sh` (no arguments), I get:
```
The IGV workspace in which the imported graphs are saved saved can be set as follows :
  1) export IGV_WORKSPACE=path/to/workspace
  2) ./igv.sh path/to/workspace
  3) default is ./workspace

IGV workspace : /home/rocastan/git/views/JDK-8324950/open/src/utils/IdealGraphVisualizer/workspace
```
However, the actual workspace directory seems to be:
`/home/rocastan/git/views/JDK-8324950/open/src/utils/IdealGraphVisualizer/application/target/userdir`

Screenshot from 2024-02-16 14-32-29

**Comment by tobiasholenstein** [comment]
> igv.sh

You are right. There was a small mistake in the `igv.sh` script. It should work now

**Comment by tobiasholenstein** [comment]
> Nice idea! When I start IGV with `sh igv.sh` (no arguments), I get:
> 
> ```
> The IGV workspace in which the imported graphs are saved saved can be set as follows :
>   1) export IGV_WORKSPACE=path/to/workspace
>   2) ./igv.sh path/to/workspace
>   3) default is ./workspace
> 
> IGV workspace : /home/rocastan/git/views/JDK-8324950/open/src/utils/IdealGraphVisualizer/workspace
> ```
> 
> However, the actual workspace directory seems to be: `/home/rocastan/git/views/JDK-8324950/open/src/utils/IdealGraphVisualizer/application/target/userdir`
> 
> Screenshot from 2024-02-16 14-32-29

You are right. There was a small mistake in the igv.sh script. It should work now

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java [hunk 2]
```diff
@@ -270,16 +317,268 @@ public void changed(InputGraphProvider lastProvider) {
...
+
+    private String workspacePath;
+
+    private void setWorkspacePath(String path) {
+        changeWorkspaceButton.setText(path);
+        workspacePath = path;
+    }
+
+    static public String WORKSPACE_XML_FILE = "graphs.xml";
+    static public String WORKSPACE_STATE_FILE = "state.igv";
```
Comment (turbanoff) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:351]: Let's use blessed modifiers order
```suggestion
    public static String WORKSPACE_XML_FILE = "graphs.xml";
    public static String WORKSPACE_STATE_FILE = "state.igv";
```

**Comment by robcasloz** [comment]
Hi Toby, this feature seems very useful, thanks for developing it! I very much prefer the current model that saves the state into the graph XML file compared to the original proposal. A few comments (this is not a full review yet):

- The export/import icons I see differ from those shown in the PR description:

outline

- I find the distinction between 'Save' and 'Export' a bit unclear from a user perspective. My suggestion would be to either remove 'Export' (one can achieve the same result by removing the unnecessary graphs and then saving all remaining ones) or rename 'Save', 'Save as...', and 'Export' with 'Save all', 'Save all as...', and 'Save selected as...' (and using similar icons for the three), if that is the only difference between saving and exporting.

- It would be nice to make more tab-specific state persistent, for example which view is displayed (sea of nodes, CFG, etc.) or whether neighboring nodes of extracted nodes are shown. This could be done here or in a future RFE.

**Comment by tobiasholenstein** [comment]
> Hi Toby, this feature seems very useful, thanks for developing it! I very much prefer the current model that saves the state into the graph XML file compared to the original proposal. A few comments (this is not a full review yet):
Thanks for the comments!
> * The export/import icons I see differ from those shown in the PR description:
> 
> outline

Strange. How did you checkout the PR? Seems like the newly added files `export.png` and `open.png` can not be found. (E.g. when applying as a patch like `git apply --index 8324950.diff` binary (png) files are not added) - Can you try:

_Checkout this PR locally:_
`$ git fetch https://git.openjdk.org/jdk.git pull/17630/head:pull/17630`
`$ git checkout pull/17630`

> * I find the distinction between 'Save' and 'Export' a bit unclear from a user perspective. My suggestion would be to either remove 'Export' (one can achieve the same result by removing the unnecessary graphs and then saving all remaining ones) or rename 'Save', 'Save as...', and 'Export' with 'Save all', 'Save all as...', and 'Save selected as...' (and using similar icons for the three), if that is the only difference between saving and exporting.
Right, I understand that it can be a bit unclear. I would suggest to drop the `Export` option and move `Import` next after `Open`:
bar

> * It would be nice to make more tab-specific state persistent, for example which view is displayed (sea of nodes, CFG, etc.) or whether neighboring nodes of extracted nodes are shown. This could be done here or in a future RFE.
Yes, this would be nice. Also IGV could reopen the last opened XML at startup. But I suggest a future RFE for this.

**Comment by robcasloz** [comment]
> Strange. How did you checkout the PR? Seems like the newly added files export.png and open.png can not be found. (E.g. when applying as a patch like git apply --index 8324950.diff binary (png) files are not added) - Can you try:

My bad, I did check the PR out via a plain-text diff file.

> I would suggest to drop the Export option and move Import next after

Sounds good to me.

> Yes, this would be nice. Also IGV could reopen the last opened XML at startup. But I suggest a future RFE for this.

Fair enough :)

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/actions/RemoveAllAction.java [hunk 3]
```diff
@@ -42,7 +43,8 @@ public String getName() {
     }
 
     public RemoveAllAction() {
-        putValue(Action.SHORT_DESCRIPTION, "Remove all graphs and groups");
+        putValue(Action.SHORT_DESCRIPTION, "clear workspace");
```
Comment (robcasloz) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/actions/RemoveAllAction.java:46]:
```suggestion
        putValue(Action.SHORT_DESCRIPTION, "Clear workspace");
```

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/resources/com/sun/hotspot/igv/coordinator/actions/Bundle.properties [hunk 4]
```diff
@@ -1,8 +1,9 @@
-CTL_ImportAction=Open...
+CTL_ImportAction=Import...
+CTL_OpenAction=Open...
 CTL_DiffGraphAction=Difference to current graph
 CTL_RemoveAction=Remove selected graphs and groups
 CTL_RemoveAllAction=Remove all graphs and groups
```
Comment (robcasloz) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/resources/com/sun/hotspot/igv/coordinator/actions/Bundle.properties:5]:
```suggestion
CTL_RemoveAllAction=Clear workspace
```

**Review by robcasloz — COMMENTED** [review]
Looks better, thanks! A couple of suggestions, for consistency between the File menu and the Outline icons.

**Review by robcasloz — COMMENTED** [review]
When a graph file with several graphs is opened (e.g. [three-graphs.zip](https://github.com/openjdk/jdk/files/14932810/three-graphs.zip)), clicking on the tabs of the opened graphs does not highlight the corresponding entries in the Outline tree (as done when the graphs are freshly imported). Would it be possible to have the same behavior regardless of whether the displayed graphs are opened or imported?

**Comment by tobiasholenstein** [comment]
> When a graph file with several graphs is opened (e.g. [three-graphs.zip](https://github.com/openjdk/jdk/files/14932810/three-graphs.zip)), clicking on the tabs of the opened graphs does not highlight the corresponding entries in the Outline tree (as done when the graphs are freshly imported). Would it be possible to have the same behavior regardless of whether the displayed graphs are opened or imported?

It should be fixed now. The entries in the Outline tree are computed lazily. So they don't exist until the folder is extended

**Comment by tobiasholenstein** [comment]
I updated the PR: save the visibleNodes instead of the hiddenNodes. This is way more memory efficient. Especially for large graphs. 

PR is ready for review again!

**Comment by robcasloz** [comment]
> It should be fixed now. The entries in the Outline tree are computed lazily. So they don't exist until the folder is extended

Thanks, that works for me now!

> I updated the PR: save the visibleNodes instead of the hiddenNodes. This is way more memory efficient. Especially for large graphs.

One issue with that is that now graphs are opened with a fixed set of visible nodes rather than "all nodes are visible". This works for the graph that is directly shown when the file is open, but if you step into the next graph in the group, new nodes will be hidden. To reproduce this, you can open [one-graph.zip](https://github.com/openjdk/jdk/files/14943396/one-graph.zip) and click on "Show next graph of current group". In the next graph, nodes 5, 7, 25, 26, and 27 are hidden unexpectedly.

One (hopefully simple) fix would be to show all nodes if `visibleNodes` is absent from the XML file or introduce an attribute to indicate that all nodes should be visible, something like:
```
<visibleNodes all="true">
</visibleNodes>
```
This would also lead to more compact XML files for graphs without hidden nodes.

Another solution would be to express the graph state as a chain of actions (e.g. `<showNode id="2"/>`, `<hideNode id="5"/>`, etc.). Then one could determine which chain of actions is more compact for a given graph state. This would also open for sending "commands" to IGV in the future, e.g. via gdb. But this is probably best left for future work.

**Comment by tobiasholenstein** [comment]
> One issue with that is that now graphs are opened with a fixed set of visible nodes rather than "all nodes are visible". This works for the graph that is directly shown when the file is open, but if you step into the next graph in the group, new nodes will be hidden. To reproduce this, you can open [one-graph.zip](https://github.com/openjdk/jdk/files/14943396/one-graph.zip) and click on "Show next graph of current group". In the next graph, nodes 5, 7, 25, 26, and 27 are hidden unexpectedly.

I opted for your proposal introducing an attribute that indicates if all nodes should be visible:

- Show all nodes an ignore nodes in <visibleNodes> (should be empty anyways)
```xml
<visibleNodes all="true">
</visibleNodes>
```

- Show no nodes
```xml
<visibleNodes all="false">
</visibleNodes>
```

- Show only nodes "0"
```xml
<visibleNodes all="false">
<node id="0"/>
</visibleNodes>
```

**Comment by robcasloz** [comment]
Thanks for addressing my questions and comments so far, Toby!

One more high-level question: currently, it seems the graph state is restored only for graphs from opened graph files, but not for graphs that are imported (e.g. from other graph files or from the network). Would it be hard to restore also the state from imported graphs? This is probably not a high-priority use case, but would be nice to e.g. enable sending graphs from gdb that are opened directly and where e.g. one node is highlighted.

**Comment by tobiasholenstein** [comment]
> Thanks for addressing my questions and comments so far, Toby!
> 
> One more high-level question: currently, it seems the graph state is restored only for graphs from opened graph files, but not for graphs that are imported (e.g. from other graph files or from the network). Would it be hard to restore also the state from imported graphs? This is probably not a high-priority use case, but would be nice to e.g. enable sending graphs from gdb that are opened directly and where e.g. one node is highlighted.

If graphs are sent over the network with `<graphStates>` they are automatically opened. But this feature is currently not available in C2. Regarding manually importing a graph I prefer to not automatically open the Tabs.

**Comment by robcasloz** [comment]
> If graphs are sent over the network with <graphStates> they are automatically opened. But this feature is currently not available in C2. Regarding manually importing a graph I prefer to not automatically open the Tabs.

I see, thanks for the clarification. I am fine with this decision, the main use case I see is for when graphs are imported over the network.

**Review by robcasloz — CHANGES_REQUESTED** [review]
There is an issue with saved difference graph states. If I open [diff.zip](https://github.com/openjdk/jdk/files/14976481/diff.zip) (which I just created by importing some graphs, opening one of them, and diffing it against another one), I get the following assertion error:
```
[INFO] java.lang.AssertionError
[INFO] 	at com.sun.hotspot.igv.util.RangeSliderModel.setPositions(RangeSliderModel.java:101)
[INFO] 	at com.sun.hotspot.igv.coordinator.OutlineTopComponent.lambda$loadContext$2(OutlineTopComponent.java:481)
[INFO] 	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
[INFO] 	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
[INFO] 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
[INFO] 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
[INFO] 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
[INFO] 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
[INFO] 	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
[INFO] 	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
[INFO] [catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
[INFO] 	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```
I imagine fully supporting saving and restoring the diff state would require quite a lot of additional complexity, both in IGV and in the XML files. Maybe this is not a very important use case, and we could just not support it?

**Comment by tobiasholenstein** [comment]
> There is an issue with saved difference graph states. If I open [diff.zip](https://github.com/openjdk/jdk/files/14976481/diff.zip) (which I just created by importing some graphs, opening one of them, and diffing it against another one), I get the following assertion error:
> 
> ```
> [INFO] java.lang.AssertionError
> [INFO] 	at com.sun.hotspot.igv.util.RangeSliderModel.setPositions(RangeSliderModel.java:101)
> [INFO] 	at com.sun.hotspot.igv.coordinator.OutlineTopComponent.lambda$loadContext$2(OutlineTopComponent.java:481)
> [INFO] 	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
> [INFO] 	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:773)
> [INFO] 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:720)
> [INFO] 	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:714)
> [INFO] 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
> [INFO] 	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
> [INFO] 	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
> [INFO] 	at org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
> [INFO] [catch] at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
> [INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
> [INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
> [INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
> [INFO] 	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
> [INFO] 	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
> ```
> 
> I imagine fully supporting saving and restoring the diff state would require quite a lot of additional complexity, both in IGV and in the XML files. Maybe this is not a very important use case, and we could just not support it?

Thanks for catching that! I fixed it. Difference graph are supported as long as they are in the same group. A difference graph from two different groups is just not saved to the xml.

**Comment by robcasloz** [comment]
> Thanks for catching that! I fixed it. Difference graph are supported as long as they are in the same group. A difference graph from two different groups is just not saved to the xml.

Great, that works fine now, as far as I can see. A minor related issue is that difference graphs from two different groups are not closed when the workspace is cleared. Is this expected? I would intuitively expect them to be closed as well, but I guess one can also argue that they do not belong to the workspace.

**Comment by tobiasholenstein** [comment]
> > Thanks for catching that! I fixed it. Difference graph are supported as long as they are in the same group. A difference graph from two different groups is just not saved to the xml.
> 
> Great, that works fine now, as far as I can see. A minor related issue is that difference graphs from two different groups are not closed when the workspace is cleared. Is this expected? I would intuitively expect them to be closed as well, but I guess one can also argue that they do not belong to the workspace.

Right. Implementation wise difference graphs from two groups are kind of a special case. But I agree the user would expect all tabs to be closed. Since we only ever have one workspace at the time, we can simply close all tabs when the workspace is closed - that's what I added in the latest commit. I think this is a good enough solution

**Review by robcasloz — APPROVED** [review]
Looks good!

**Review by robcasloz — APPROVED** [review]

**Comment by chhagedorn** [comment]
That is a nice feature! I've tried the following out which does not seem to be working as expected:

1) Open graphs.xml -> opens with extracted nodes as shown in PR description
2) Click "Clear workspace"
3) Open graphs.xml again -> this time, it opens the entire graph without selection.

Same result when I completely restart IGV in step 2) instead.

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java [hunk 5]
```diff
@@ -271,15 +321,233 @@ public void changed(InputGraphProvider lastProvider) {
...
     @Override
-    public void readExternal(ObjectInput objectInput) throws IOException, ClassNotFoundException {
-        // Not called when user starts application for the first time
-        super.readExternal(objectInput);
-        ((BeanTreeView) this.treeView).setRootVisible(false);
+    public boolean canClose() {
+        SwingUtilities.invokeLater(() -> {
+            this.clearWorkspace();
+            this.open(); // Reopen the OutlineTopComponent
+            this.requestActive();
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:328]: `this` can probably be removed
```suggestion
            clearWorkspace();
            open(); // Reopen the OutlineTopComponent
            requestActive();
```

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java [hunk 6]
```diff
@@ -271,15 +321,233 @@ public void changed(InputGraphProvider lastProvider) {
...
+    private boolean overwriteDialog(String filename) {
+        JFrame frame = new JFrame();
+        String message = "Do you want to overwrite " + filename + "?";
+        int result = JOptionPane.showConfirmDialog(frame, message, "Confirm Overwrite", JOptionPane.YES_NO_OPTION, JOptionPane.WARNING_MESSAGE);
+        if (result == JOptionPane.YES_OPTION) {
+            frame.dispose();
+            return true;
+        }
+        frame.dispose();
+        return false;
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:388]: Could be simplified to:
```
frame.dispose();
return result == JOptionPane.YES_OPTION;
```

File: src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java [hunk 7]
```diff
@@ -271,15 +321,233 @@ public void changed(InputGraphProvider lastProvider) {
...
+    }
+
+    /**
+     * Loads a graph document from the given file path, updating progress via a ProgressHandle.
+     * Parse the XML file, add the parsed document to the workspace, and load associated contexts if specified.
+     */
+    private void loadGraphDocument(String path, boolean loadContext) throws IOException {
+        if (path == null || Files.notExists(Path.of(path))) {
+            return;
+        }
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:499]: Can `path` really be `null` here?

File: src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/serialization/Parser.java [hunk 8]
```diff
@@ -367,118 +308,164 @@ protected InputNode start() throws SAXException {
...
+                    }
+                }
+
+                if (!blocksContainNodes) {
+                    graph.clearBlocks();
+                    blockConnections.clear();
                 } else {
-                    addStarted.run();
+                    // Blocks and their nodes defined: add other nodes to an
+                    //  artificial "no block" block
```
Comment (chhagedorn) [inline src/utils/IdealGraphVisualizer/Data/src/main/java/com/sun/hotspot/igv/data/serialization/Parser.java:405]:
```suggestion
                    // artificial "no block" block
```

**Review by chhagedorn — APPROVED** [review]
As discussed offline, the confusion was due to opening the tabs in a random order each time a file is opened. The fix now looks good.

Only a few minor comments - I'm not familiar with the code. I've dong some testing with IGV and the features seem to work. Great job!

**Review by chhagedorn — APPROVED** [review]
Looks good, thanks for the update!

**Comment by tobiasholenstein** [comment]
Thanks for the reviews @robcasloz and @chhagedorn

/integrate

**Pull request closed without merging**